openstack: Remove domain name when trying to resolve a hostname.

jsafrane opened the pull request:
Depending on OpenStack deployment, hostname of a virtual machine (a.k.a server)
is either directly the machine [instanceName] or [instanceName].[domain].

Now, when looking for appropriate OpenStack object representing a virtual
machine try to remove the domain component from hostname if search for full
hostname fails.

jsafrane commented:
Resolves #13556

jsafrane commented:
cc @spothanis, what's your opinion?

thockin commented:
@spothanis @anguslees @ashw7n help?

spothanis commented:
Sorry for the delayed reply. 
For #13556 I was thinking if we could pass-in the `instance-id` to the cinder volume plugin, that `kubelet` is already aware of, then we won't have to match the hostnames to the server names and discover the server. But from my cursor glance, it looks like `volumehost` is not aware of the name of the node, so it might be slightly more involved change.

However, it looks ok to match server names in the way proposed here. Also, I am sure you must already know that, getServerByName is used in node registration and heartbeat.

anguslees commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -234,6 +234,32 @@ func (i *Instances) List(name_filter string) ([]string, error) {
+	// the server ID does not have it.
+
+	// Try with the full name first
+	server, err := getServerByName(client, hostname)
```
Does it work if we look up by ^shortname(\.suffix)?$ rather than perform multiple lookups?

jsafrane replied:
It does, at least in my openstack installation. From my logs:
```
Looking for server with ID ^jsafrane-k2(\.censored\.domain\.name)?$
Found server jsafrane-k2
```

anguslees commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -881,8 +908,9 @@ func (os *OpenStack) getVolume(diskName string) (volumes.Volume, error) {
 
+// getComputeIDbyHostname tries to find ID of the OpenStack instance where
+// this code runs.
 func (os *OpenStack) getComputeIDbyHostname(cClient *gophercloud.ServiceClient) (string, error) {
```
I know this isn't part of your change, but this entire function confuses me.  Why do we need to fetch *all* instances and then matching hostnames on the client side?

I would have expected this function to be something as simple as (with a few more lines of error handling):
```
   hostname, err := ossys.Hostname()
   i, ok := os.Instances()
   return i.InstanceID(hostname)
```

jsafrane replied:
@spothanis, I think this is your code, what's your opinion? I am in favor of replacing it with simple `getServerByHostName(ossys.Hostname())` call.

spothanis replied:
I had originally written it that way because server name can be different from hostname.
For ex in our env: Name is `kubernetes-master-3`, hostname is `kubernetes-master-3-6050`. 

Later I added the name check to it cause I think @jsafrane you said that hostname in your env was empty however your server names matched the hostnames.

anguslees replied:
Ah... So I _think_ the code below no longer works with your `kubernetes-master-3-6050` example since I don't see anything that ignores an arbitrary suffix.  In general, the hostname might have nothing at all to do with the openstack instance name, so if the goal is just to find the instance ID of the current host, then it would be good to explore other approaches entirely.

Perhaps we should just read `/var/lib/cloud/data/instance-id` and/or do the same config-drive or metadata lookup  ourselves?

The code elsewhere in this current change looks fine from a code pov, but I'm beginning to feel the original motivation for it is perhaps not the best direction to pursue.  Would reading instance-id from config-drive / metadata server address the original need?

spothanis replied:
I completely agree that the hostname matching ugly-ness should be removed. `instance-id` is already discovered by the kubelet during initial node [registration](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L872). I wanted to pass this `instance-id` from cinder volume plugin directly to the detach function.  Though `volume-host` is cloud aware it is not does not know the nodename. 

I did explore the config drive way, but I wasn't sure if it is ok to assume that it will be present everywhere. It would definitely address the original need. We could look at both, config drive and metadata server, if that works.

jsafrane replied:
What about this:

```golang
id, err := ioutil.ReadFile("/var/lib/cloud/data/instance-id")
if err == nil {
    return id
}
// try hostname if /var/lib/cloud is not available
hostname, err := ossys.Hostname()
return getServerByHostName(hostname)
```

spothanis replied:
:+1:, we can add metadata server later

jsafrane replied:
Implemented.

jsafrane commented:
I added "Review changes." patch, which does one Openstack query instead of two (+some logs around)

liggitt commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -234,9 +234,35 @@ func (i *Instances) List(name_filter string) ([]string, error) {
+	if domain == "" {
+		nameRegExp = fmt.Sprintf("^%s$", regexp.QuoteMeta(hostname))
+	} else {
+		nameRegExp = fmt.Sprintf("^%s(\\.%s)?$", regexp.QuoteMeta(shortname), regexp.QuoteMeta(domain))
```
Could a server exist with just the shortname AND one exist with the full name? What would happen in that scenario?

jsafrane replied:
In theory, yes, it could happen. IMO it's highly unlikely. When it happens, getServerByRegExp fails with ErrMultipleResults.

liggitt replied:
I still think we should handle the case. Where both exist, match the exact match

liggitt replied:
especially since that could break cases that work today, where there is an exact match for the long name

jsafrane replied:
I added some code to prefer whole hostname to shortned name without domain (+ obligatory rebase).

jsafrane commented:
I reworked the patch to use /var/lib/.../instance-id first and try hostname if it fails.

pmorie commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -234,9 +235,39 @@ func (i *Instances) List(name_filter string) ([]string, error) {
+	glog.V(5).Infof("Looking for server with ID %s", nameRegExp)
 	opts := servers.ListOpts{
-		Name:   fmt.Sprintf("^%s$", regexp.QuoteMeta(name)),
+		Name:   nameRegExp,
```
The format of the option changed here -- is that important?

pmorie replied:
Nevermind, I see that it moved up a frame.

pmorie replied:
Nit: `nameRegExp` isn't really a regular expression to evaluate, it's the name.

jsafrane replied:
well, it was regexp, but it does not matter, now it's competelly rewritten

pmorie commented:
@liggitt Can you take another pass through this one when you have a chance?

liggitt commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -261,9 +289,26 @@ func getServerByName(client *gophercloud.ServiceClient, name string) (*servers.S
+		var lastMatch *servers.Server
+		// try to find exactMatch
+		for _, server := range serverList {
+			if server.Name == exactMatch {
```
is server.Name not unique in openstack?

anguslees replied:
server.Name is a label chosen by the user at instance creation, and is not necessarily unique.

jsafrane replied:
That's quite different to GCE and maybe others. Kubernetes assume that instance name is unique and all the methods operate with names as identifiers of nodes. We could use uuid as instance name, but that would require basically to refactor whole OpenStack cloud provider.

jsafrane replied:
I checked that if there are two servers with the same Name, kubelet fails to start on them:

```
E1109 15:01:13.517363    9576 kubelet.go:929] Unable to construct api.Node object for kubelet: failed to get external ID from cloud provider: Multiple results where only one expected
```

Do we need to care about these use cases?

liggitt replied:
But you're not using the node name, you're using os.Hostname(), which may or may not match the node name

jsafrane replied:
The point was that Kubelet won't even reach my code because it fails to initialize.

liggitt commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -886,43 +932,30 @@ func (os *OpenStack) getVolume(diskName string) (volumes.Volume, error) {
+	idBytes, err := ioutil.ReadFile(instanceIDFile)
+	if err == nil {
+		id := string(idBytes)
+		id = strings.TrimSpace(id)
```
do we need to check a non-empty value was obtained?

anguslees replied:
Possibly, but "empty" is just one of several invalid values that might be in this file.  I'm sort of inclined to go with "If the file exists, trust it" without a clear reason not to...

liggitt replied:
An invalid value would at least show up in a reasonable way in errors later (if the value is "bogus", we'd see an error trying to find the "bogus" server in openstack). If this file is present, but contains no value, I think this should probably return an error

liggitt commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -886,43 +932,30 @@ func (os *OpenStack) getVolume(diskName string) (volumes.Volume, error) {
-		if err != nil {
-			return "", err
-		} else {
-			if (server.Metadata["hostname"] != nil && server.Metadata["hostname"] == hostname) || (len(server.Name) > 0 && server.Name == hostname) {
```
are we no longer matching on this? are all openstack server names guaranteed to match either the shortname or fqdn of our hostname?

anguslees replied:
No the server name is not guaranteed to match the hostname *at all* - which is why I'm trying to move this change to using the appropriate metadata apis instead..

anguslees commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -886,43 +932,30 @@ func (os *OpenStack) getVolume(diskName string) (volumes.Volume, error) {
+	}
+	glog.V(4).Infof("Reading %s failed: %v, trying hostname", instanceIDFile, err)
 
+	// try to find ID from hostname
```
Ideally we'd never need to guess instance-id from hostname.  What would it take to remove this fallback?

The other methods that might be provided in OpenStack (afaik) are:
- metadata service
- config drive

[Metadata service](http://docs.openstack.org/admin-guide-cloud/compute-networking-nova.html#metadata-service) is easy:  Just fetch http://169.254.169.254/openstack/2012-08-10/meta_data.json and return the `uuid` value

[Config drive](http://docs.openstack.org/user-guide/cli_config_drive.html) is more complicated - we need to mount the block device labelled `config-2` and then parse the same `uuid` value from `<mount point>/openstack/2012-08-10/meta_data.json`.  This obviously requires the permissions to mount filesystems - but the kubelet already has that for persistent volumes (always?).

Regardless of the method used to fetch it, we should cache value in memory since it will never change and will be used multiple times by the other functions in this file.

jsafrane replied:
> Regardless of the method used to fetch it, we should cache value in memory since it will never change and will be used multiple times by the other functions in this file.

Ok, patch will follow shortly.

> Just fetch http://169.254.169.254/openstack/2012-08-10/meta_data.json

`http://169.254.169.254/2007-01-19/meta-data/instance-id` looks easier to parse. Can it be used instead of the JSON? Is it stable enough?

jsafrane replied:
replying to myself...

> http://169.254.169.254/2007-01-19/meta-data/instance-id looks easier to parse. Can it be used instead of the JSON? Is it stable enough?

It cannot, it's not instance ID (=uuid) but some other string. *Damn*.

jsafrane commented:
3rd version of the patch. Now:
* InstanceID is used to attach/detach volumes.
* And it's retrieved from /var/lib/cloud/data/instance-id or http://169.254.169.254/2007-01-19/meta-data/instance-id on startup and cached afterwards.
* No other code than attach/detach is modified.

This patch does not fix wrong assumption that server names are unique, Kubelet still won't start. It was never working, this patch does not make it better or worse. [ I'll fix it in subsequent PR, I need to fix the volumes ASAP and not to break anything else... Basically I'll use InstanceID instead of server name everywhere.]

liggitt commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -140,17 +142,55 @@ func readConfig(config io.Reader) (Config, error) {
+	glog.V(5).Infof("Cannot read %s: '%v', trying metadata server", instanceIDFile, err)
+
+	const instanceIDUrl = "http://169.254.169.254/2007-01-19/meta-data/instance-id"
+	resp, err := http.Get(instanceIDUrl)
```
should this operation have a timeout? http.DefaultClient has no timeout...

liggitt replied:
actually, I guess it uses http.DefaultTransport, which has a 30 second timeout... if that errors, it'll happen once, right, and the newOpenStack() call will fail, so we wouldn't hit that error repeatedly

liggitt commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -140,17 +142,55 @@ func readConfig(config io.Reader) (Config, error) {
+		glog.V(3).Infof("Cannot read %s: %v", instanceIDUrl, err)
+		return "", err
+	}
+
```
check `resp.StatusCode` is as expected? err can be nil for non-200 responses

liggitt replied:
and maybe content-type? (I don't know the API well enough to know what is expected)

jsafrane replied:
I added some checks

liggitt commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -140,17 +142,55 @@ func readConfig(config io.Reader) (Config, error) {
 		return nil, err
 	}
 
+	id, err := readInstanceID()
```
currently, `readInstanceID` can return empty strings... either return errors in that case, or check for empty instance id here

jsafrane replied:
Ok

liggitt commented:
This approach seems a lot better, just a few nits around the metadata request handling and empty instance ID error cases

jsafrane commented:
I was wrong about `http://169.254.169.254/2007-01-19/meta-data/instance-id`, it does not contain instance UUID. I added some ugly json parsing.

liggitt commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -140,17 +143,113 @@ func readConfig(config io.Reader) (Config, error) {
+	glog.V(5).Infof("Cannot read %s: '%v', trying metadata server", instanceIDFile, err)
+
+	// Try to get JSON from metdata server.
+	const metadataUrl = "http://169.254.169.254/openstack/2012-08-10//meta_data.json"
```
typo on `//`?

jsafrane replied:
Good catch!

liggitt commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -140,17 +143,113 @@ func readConfig(config io.Reader) (Config, error) {
+// parseMetadataUUID reads JSON from OpenStack metadata server and parses
+// instance ID out of it.
+func parseMetadata(jsonData []byte) (string, error) {
+	var jsonif interface{}
```
jsonif := map[string]interface{}{}

liggitt commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -140,17 +143,113 @@ func readConfig(config io.Reader) (Config, error) {
+
+	// We should receive an object with { 'uuid': '<uuid> } and couple of other
+	// properties (which we ignore).
+	obj, ok := jsonif.(map[string]interface{})
```
making jsonif typed lets you drop this cast and check

jsafrane replied:
Done

liggitt commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -140,17 +143,113 @@ func readConfig(config io.Reader) (Config, error) {
+	// OpenStack 'Kilo' sends 'text/html; charset=UTF-8', be prepared for text/plain and no
+	// conent-type at all.
+	cType := resp.Header.Get("Content-Type")
+	if cType != "" && !strings.HasPrefix(cType, "text/plain") && !strings.HasPrefix(cType, "text/html") {
```
drop the content type check... I only wanted it if it was sane... the returned content type has nothing to do with the actual content (for example, this would break if the server ever started returning "application/json")

jsafrane replied:
Done

liggitt commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -140,17 +143,113 @@ func readConfig(config io.Reader) (Config, error) {
 
+// parseMetadataUUID reads JSON from OpenStack metadata server and parses
+// instance ID out of it.
+func parseMetadata(jsonData []byte) (string, error) {
```
parseMetadataUUID is a better name for the function

jsafrane replied:
Renamed

liggitt commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -140,17 +143,113 @@ func readConfig(config io.Reader) (Config, error) {
+		glog.V(3).Infof("Cannot parse instance ID from metadata from %s: %v", metadataUrl, err)
+		return "", err
+	}
+	instanceID = strings.TrimSpace(instanceID)
```
trimming the value read from structured JSON seems bad... also, I would expect any manipulation of the uuid (trimming or comparing to "") to happen inside `parseMetadataUUID`

jsafrane replied:
Done

liggitt commented:
parsing and fallback logic looks good to me. @anguslees, can you double check the metadata query? The only other question I have is if the metadata query would from inside a containerized kubelet

jsafrane commented:
> GCE e2e build/test failed for commit 72d29af.

Is it a flake?
```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:911
  Kubectl describe
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:560
    should check if kubectl describe prints relevant information for rc and pods [Conformance] [It]
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:559

    Nov  9 10:23:50.767: Failed to find Events: in Name:		redis-master
    Namespace:	e2e-tests-kubectl-0979r
    Image(s):	redis
    Selector:	app=redis,role=master
    Labels:		app=redis,role=master
    Replicas:	1 current / 1 desired
    Pods Status:	1 Running / 0 Waiting / 0 Succeeded / 0 Failed
    No volumes.
    No events.

    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:922
```

anguslees commented:
I wanted to help progress this change, so I wrote my own version at https://github.com/kubernetes/kubernetes/pull/17041
- it seems @jsafrane beat me to it with his latest update (and I haven't tested mine) - so feel free to steal any or no bits ;)

anguslees commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -140,17 +143,101 @@ func readConfig(config io.Reader) (Config, error) {
+	// We should receive an object with { 'uuid': '<uuid> } and couple of other
+	// properties (which we ignore).
+
+	var obj map[string]interface{}
```
Fyi, this can be done more easily (and idiomatically) by Unmarshalling into a struct - see https://github.com/kubernetes/kubernetes/pull/17041/files#diff-9223cd6a0f3d50c7e3bd990b85d1c09fR41

jsafrane replied:
I know, I just felt that `type Metadata struct {uuid string}` is a bit stupid.

liggitt replied:
if the type really bothers you, you can do it inline, like this:
```
obj := struct{ UUID string }{}
json.Unmarshal(jsonData, &obj)
```

pmorie replied:
+ @liggitt

jsafrane replied:
Ok, I reworked parsing to

```go
	obj := struct{ UUID string }{}
	err := json.Unmarshal(jsonData, &obj)
	if err != nil {
		return "", err
	}

	uuid := obj.UUID
	if uuid == "" {
		err = fmt.Errorf("cannot parse OpenStack metadata, got empty uuid")
		return "", err
	}

	return uuid, nil
```

anguslees commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -140,17 +143,101 @@ func readConfig(config io.Reader) (Config, error) {
+		}
+		// Fall through with empty instanceID and try metadata server.
+	}
+	glog.V(5).Infof("Cannot read %s: '%v', trying metadata server", instanceIDFile, err)
```
Many openstack installs don't offer metadata service and use config drive instead.  If you'd like to support that (without cloud-init), I have some (untested) code in https://github.com/kubernetes/kubernetes/pull/17041/files#diff-9223cd6a0f3d50c7e3bd990b85d1c09fR50

pmorie replied:
@anguslees Are you interested in doing a follow-up to add that?

anguslees commented:
@jsafrane Thanks, this looks much more robust than the current logic :smiley:   Assuming you've tested it against a real metadata service, this looks good to me.  See my other change if you'd like to also include support for config drive, otherwise we can add that in a separate follow-up change.

@liggitt Interesting question re kubelet in a container.  I think it will require some sort of elevated container privileges:
- Reading `/var/lib/cloud/data/instance-id` will obviously fail unless that is bind-mounted in from outside (easy enough)
- Mounting config-drive will fail unless the raw device is bind-mounted in from outside (and I think cgroups might prevent device access - perhaps we need a privileged container for this?)
- Accessing 169.254.169.254 from inside the container will work if the outer host forwards that out the regular eth0 - which I think _most_ setups should do.  Obviously net=host makes that easier.

Incidentally, the ease with which 169.254.169.254 may be accessed by just about any process on the VM is one of the reasons many deployments opt to disable it and instead rely on config drive.

jsafrane commented:
@anguslees, thanks for the review and your comments, I'll look at config drive in separate PR, when this one is merged. Also, I'll try to make Kubelet running on servers, which have duplicate names.

anguslees commented:
(Aside: The current behaviour is that openstack server names are assumed (and verified) to be unique (amongst ACTIVE servers).  This is the 'name' by which k8s knows the minion, and then there's various provider methods added since that fetch various forms of poorly defined "provider IDs" from that name.  This made sense back when there was a k8s command line option that allowed you to automatically find minions with provider names that matched a regex - but it looks like that has been removed(?) and I don't think Instances.List() is even called from anywhere anymore.  There's no problem with the current behaviour, but it might be nice to switch to uuids across the board and reduce a few unnecessary lookups)

jsafrane commented:
Yes, I think `InstancesList()` is not used any more. Still, there is `.TCPLoadBalancer.EnsureTCPLoadBalancer(..., hosts []string, ...)`, where `hosts` are *probably* node names.

liggitt commented:
I'm not sure I would expect the kubelet to mount config drive... I could see it using it if already available

pmorie commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -140,17 +143,101 @@ func readConfig(config io.Reader) (Config, error) {
+	glog.V(5).Infof("Cannot read %s: '%v', trying metadata server", instanceIDFile, err)
+
+	// Try to get JSON from metdata server.
+	const metadataUrl = "http://169.254.169.254/openstack/2012-08-10/meta_data.json"
```
Is there a reason for hard-coding to an IP?

jsafrane replied:
Yes, that's the only IP address where metadata service can be found. http://docs.openstack.org/admin-guide-cloud/compute-networking-nova.html#metadata-service

pmorie replied:
Okay, can you add an explanatory comment for that?  It looks seriously weird/bad due to all my previous experience.  It should also move out of this method into package scope

jsafrane replied:
Done, `metadataUrl` is on the package level
```go
package openstack

// <imports, other constants>

// metadataUrl is URL to OpenStack metadata server. It's hadrcoded IPv4
// link-local address as documented in "OpenStack Cloud Administrator Guide",
// chapter Compute - Networking with nova-network.
// http://docs.openstack.org/admin-guide-cloud/compute-networking-nova.html#metadata-service
const metadataUrl = "http://169.254.169.254/openstack/2012-08-10/meta_data.json"
```

jsafrane committed:
openstack: cache InstanceID and use it for volume management.

Hostnames are not related to server names or instance IDs in any way,
don't use them when attaching/detaching volumes.

jsafrane commented:
Does kubelet always run in a privileged container? If so, It should see the config-2 device and mount it it  (or see it mounted somewhere).

anguslees commented:
@jsafrane: I think kubelet most often doesn't run in a container at all (even on something like coreos, aiui).   More importantly, I think you're concerned about something that isn't in this PR - this PR (currently) has no support for config drive at all ;)  I think we _should_ add support for it (in a later change), and I think it would be better to address your concerns at that time when we have something concrete in front of us.

jsafrane commented:
Ok, can we merge this PR? If anyone has issues / suggestions, we can continue somewhere else.

pmorie commented:
@jsafrane The containerized kubelet should always be run in a privileged container, yes

jsafrane commented:
> GCE e2e build/test failed for commit 7fc8bca.

Flake?

```
Summarizing 5 Failures:

[Fail] Pods [It] should have their auto-restart back-off timer reset on image update 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/pods.go:193

[Fail] Kubectl client [BeforeEach] Kubectl label should update the label on a resource [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework.go:72

[Fail] Services [It] should be able to change the type and nodeport settings of a service 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/service.go:485

[Fail] Pods [BeforeEach] should support remote command execution over websockets 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework.go:72

[Fail] Kubectl client [BeforeEach] Update Demo should do a rolling update of a replication controller [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework.go:72

```

pmorie commented:
@k8s-bot test this

swagiaal commented:
@k8s-bot test this

wojtek-t commented:
@k8s-bot e2e test this please
